primitives/transaction/src/historic_transaction.rs

```diff
@@ -159,15 +159,37 @@ impl HistoricTransaction {
         }
     }
 
-    /// Returns the hash of the underlying transaction/inherent. For reward inherents we return the
-    /// hash of the corresponding reward transaction. This results into an unique hash for the
-    /// reward inherents (which wouldn't happen otherwise) and allows front-end to fetch rewards by
-    /// their transaction hash.
+    /// Returns the hash based on the underlying transaction/event.
     pub fn tx_hash(&self) -> Blake2bHash {
         match &self.data {
             HistoricTransactionData::Basic(tx) => tx.hash(),
-            // TODO: check for freedom of hash prefixes
-            _ => self.data.hash(),
+            _ => {
+                let block_height = if matches!(self.data, HistoricTransactionData::Equivocation(_))
+                {
+                    // Equivocations are not distinguished by block height.
+                    0
+                } else {
+                    self.block_number
+                };
+                let mut hasher = Blake2bHasher::new();
+                // Ensure that the data here doesn't form a valid transaction
+                // by adhering to the same serialization structure as
+                // `Transaction::serialize_content`.
+                hasher.write_all(b"\x00\x00").unwrap(); // recipient data len
```

Comment:
I didn't understand what this comment meant until Philipp explained it to me. I would reword more like

> This ensures a separate hashing-namespace for inherents, by preventing `self.data` from serializing to a valid transaction (accidentally or on purpose), by following the serialization of a transaction until an impossible byte (sender type).

Reply:
However, as Philipp also mentioned, tx serialization might change to include execution result, and there is no compiler link between that and the manual bytes here. So how do we ensure that if the serialization changes, this code here is also updated and not forgotten?

A comment at the tx serialization?

Reply:
Did both.

Reply:
Couldn't we actually implement something that puts zeroes based on the expected types. I'm thinking on something like this:
```
write_all(0_u8 * sizeo_of(Address))
```